File: tools/serverless/src/protoclean.ts
import {Stats} from "fs";

const fs = require("fs");
const path = require("path");

function main() {
    let inputpath = '../../../proto'
    let outputpath = '../../../proto_std'
    const argv = require('yargs').argv;
    if (argv.i) inputpath= argv.i
    if (argv.o) outputpath= argv.o
    clean(inputpath,outputpath)
}

function clean(inputpath: string, outputpath: string) {
    console.log(inputpath, "[", outputpath, "]")
    if (!fs.existsSync(outputpath)) fs.mkdirSync(outputpath)
    const files = fs.readdirSync(inputpath)
    for (let file of files) {
        const newPath = path.join(inputpath, file)
        fs.stat(newPath, function (err: NodeJS.ErrnoException | null, stats: Stats) {
            if (stats.isFile()) replace(newPath, path.join(outputpath, file))
            else if (newPath.endsWith(`utils/proto`)) return
            else clean(newPath, path.join(outputpath, file))
        })
    }
}

function replace(inputfile: string, outputfile: string) {
    if (!inputfile.endsWith(".proto")) return
/*    const buf = Buffer.alloc(1024);
    fs.open(inputfile, "r+", function (err, fd) {
        if (err) return console.error(err);
        console.log("文件打开成功！");
        console.log("准备读取文件！");
        fs.read(fd, buf, 0, buf.length, 0, function (err, bytes) {
            if (err) console.log(err);
            console.log(bytes + "  字节被读取");

            // 仅输出读取的字节
            if (bytes > 0) {
                console.log(buf.slice(0, bytes).toString());
            }
        });
        // 关闭文件
        fs.close(fd, function (err) {
            if (err) console.log(err);
            console.log("文件关闭成功");
        });
    })*/
    let data = fs.readFileSync(inputfile,'utf-8')
    data = data.replace(/import "github.*\n/,'')
    data = data.replace(/import "protoc-gen-openapiv2.*\n/,'')
    data = data.replace(/import "utils\/proto\/gogo\/enum.proto.*\n/,'')
    data = data.replace(/import "google\/api\/annotations.proto.*\n/,'')
    data = data.replace(/\[\([\w.]*\) = [\[\]\\\n\w="',:;*@(){} \-\u4e00-\u9fa5\uff0c]*]/,'')
    data = data.replace(/option \[\([\w.]*\) = [\[\]\\\n\w="',:;*@(){} \-\u4e00-\u9fa5]*;/,'')
    fs.writeFileSync(outputfile,data)
}

main()

File: tools/python/protoclean.py
import os
import re
import sys, getopt


def clean(inputpath, outputpath):
    print(inputpath, "[", outputpath, "]")
    if not os.path.exists(outputpath):
        os.makedirs(outputpath)
    files = os.listdir(inputpath)
    for file in files:
        path = os.path.join(inputpath, file)
        if os.path.isfile(path):
            replace(path, os.path.join(outputpath, file))
        else:
            clean(path, os.path.join(outputpath, file))


def replace(inputfile, outputfile):
    if not inputfile.endswith(".proto"):
        return
    f = open(inputfile, "r", encoding="utf-8")
    data = f.read()
    f.close()
    data = re.sub(r"import \"github.*\n", "", data)
    data = re.sub(r"import \"protoc-gen-openapiv2.*\n", "", data)
    data = re.sub(r"import \"utils/proto/gogo/enum.proto.*\n", "", data)
    data = re.sub(r"import \"utils/proto/gogo/graphql.proto.*\n", "", data)
    data = re.sub(r"import \"google/api/annotations.proto.*\n", "", data)
    data = re.sub(r"\[\([\w\s.)=\"/:\\;\-(\'{},一-龥，]*]", "", data)
    data = re.sub(r"option \([\w\s.)/\[\]=\":*@\-('{},一-龥]*;", "", data)
    f = open(outputfile, "w+", encoding="utf-8")
    f.write(data)
    f.close()


def main(argv):
    inputpath = '../../proto'
    outputpath = '../../proto_std'

    opts, args = getopt.getopt(argv, "hi:o:", ["ipath=", "opath="])

    for opt, arg in opts:
        if opt == '-h':
            print('test.py -i <inputpath> -o <outputpath>')
            sys.exit()
        elif opt in ("-i", "--ipath"):
            inputpath = arg
        elif opt in ("-o", "--opath"):
            outputpath = arg
    print('输入的路径为：', inputpath)
    print('输出的路径为：', outputpath)
    clean(inputpath, outputpath)


if __name__ == "__main__":
    main(sys.argv[1:])
